source/tests/llassetgen-tests/Packing.cpp

```diff
@@ -0,0 +1,112 @@
+#include <gmock/gmock.h>
+
+#include <llassetgen/llassetgen.h>
+
+using llassetgen::Packing;
+
+using Vec = llassetgen::Vec2<llassetgen::PackingSizeType>;
+
+/**
+ * Base class for all packing test fixtures.
+ *
+ * Defines several test methods that should be run for all packing methods.
+ */
+class PackingTest : public testing::Test {
+   protected:
+    virtual bool run(const std::vector<Vec>& rectSizes, bool allowRotations, Packing& packing) = 0;
+    virtual Packing run(const std::vector<Vec>& rectSizes, bool allowRotations) = 0;
+
+    static bool rotatedSizesEquals(Vec size1, Vec size2) { return size1 == size2 || size1 == Vec{size2.y, size2.x}; }
+
+    /**
+     * Expect a fixed size packing to succeed and validate the result.
+     */
+    void expectSuccessfulValidPacking(const std::vector<Vec>& rectSizes, Vec atlasSize, bool allowRotations);
+
+    /**
+     * Expects a variable size packing to return a valid result.
+     */
+    void expectValidPacking(const std::vector<Vec>& rectSizes, bool allowRotations) {
+        validatePacking(run(rectSizes, allowRotations), rectSizes, allowRotations);
+    }
+
+    /**
+     * Validate a packing given the input parameters.
+     */
+    void validatePacking(Packing packing, const std::vector<Vec>& rectSizes, bool allowRotations);
+
+    void testRejectTooWide() {
+        Packing p1{{1, 1}}, p2{{1, 1}};
+        EXPECT_FALSE(run({{2, 1}}, false, p1));
+        EXPECT_FALSE(run({{2, 1}}, true, p2));
+    }
+
+    void testRejectTooHigh() {
+        Packing p1{{1, 1}}, p2{{1, 1}};
+        EXPECT_FALSE(run({{1, 2}}, false, p1));
+        EXPECT_FALSE(run({{1, 2}}, true, p2));
+    }
+
+    void testRotateOnly() {
+        Packing p{{2, 2}};
+        EXPECT_FALSE(run({{2, 1}, {1, 2}}, false, p));
+        expectSuccessfulValidPacking({{2, 1}, {1, 2}}, {2, 2}, true);
+    }
+
+    void testAcceptAtlasSized() {
+        expectSuccessfulValidPacking({{1, 1}}, {1, 1}, false);
+        expectSuccessfulValidPacking({{1, 1}}, {1, 1}, true);
+    }
+
+    void testAcceptMultipleTiny() {
+        expectSuccessfulValidPacking({{1, 2}, {3, 4}, {5, 6}, {7, 8}}, {256, 256}, false);
+        expectSuccessfulValidPacking({{1, 2}, {3, 4}, {5, 6}, {7, 8}}, {256, 256}, true);
+    }
+
+    void testVariableSizePacking() {
+        expectValidPacking({{1, 2}, {3, 4}, {5, 6}}, false);
+        expectValidPacking({{1, 2}, {3, 4}, {5, 6}}, true);
+    }
+};
+
+void PackingTest::expectSuccessfulValidPacking(const std::vector<Vec>& rectSizes, Vec atlasSize, bool allowRotations) {
+    Packing packing{atlasSize};
+    ASSERT_TRUE(run(rectSizes, allowRotations, packing));
+    EXPECT_EQ(atlasSize, packing.atlasSize);
+    validatePacking(packing, rectSizes, allowRotations);
+}
+
+void PackingTest::validatePacking(Packing packing, const std::vector<Vec>& rectSizes, bool allowRotations) {
+    EXPECT_EQ(rectSizes.size(), packing.rects.size());
+
+    for (size_t i = 0; i < packing.rects.size(); i++) {
```

Comment:
optimization: if packing.rects.size() doesn't change while looping, you can store the size in a variable. This way,
(1) the size-function is not called for every iteration.
(2) the reader knows that it doesn't change while looping.

Reply:
I'm pretty sure almost every compiler will optimize that function call away. Even so, this is test code, so I'd prefer it to be concise and readable over perfectly optimized (and with the current tests we really shouldn't have any performance issues).